portfolio/src/main/webapp/js/background.js

```diff
@@ -0,0 +1,375 @@
+     */
+    this.height = height;
+
+    /**
+     * Width of the tank.
+     * @type {number} @const
+     */
+    this.width = width;
+
+    /**
+     * Effective radius of static effect.
+     * @type {number} @const
+     */
+    this.staticRadius = staticRadius;
+
+    /**
+     * Effective diameter of static effect.
+     * @type {number} @const
+     */
+    this.staticDiameter = staticRadius * 2;
+
+    /**
+     * The size of the static balls.
+     * @type {number} @const
+     */
+    this.ballSize = ballSize;
+
+    /**
+     * Spatial hashing grid used to reduce calls to balls
+     * beyond effective radius.
+     * @type {Object.<string, Array<Ball>>} @private
+     */
+    this._spatialHash = {};
+
+    /**
+     * Array of statically-charged balls.
+     * @type {Array<Ball>} @private
+     */
+    this._balls = [];
+  }
+
+  /**
+   * Updates the balls in the tank by one frame.
+   */
+  update() {
+    this._balls.forEach((ball, i) => {
+      // Record previous pos to update spatial hashing
+      const prevPos = createVector(ball.pos.x, ball.pos.y);
+      ball.update(this);
+      this._updateHash(prevPos, ball);
+    });
+  }
+
+  /**
+   * Draws the balls in the tank.
+   */
+  draw() {
+    this._balls.forEach((ball, i) => ball.draw(this));
+  }
+
+  /**
+   * Adds a ball to the tank.
+   * @param {Ball} ball
+   */
+  addBall(ball) {
+    this._balls.push(ball);
+    this._placeInHash(this._getSpatialHashPos(ball.pos), ball);
+  }
+
+  /**
+   * Updates the spatial hashing position of a ball.
+   * @param {p5.Vector} prevPos
+   * @param {Ball} ball
+   * @private
+   */
+  _updateHash(prevPos, ball) {
+    const prevHashPos = this._getSpatialHashPos(prevPos);
+    const currentHashPos = this._getSpatialHashPos(ball.pos);
+    if (!prevHashPos.equals(currentHashPos)) {
+      if (!this._removeFromHash(prevHashPos, ball)) {
+        throw "Invalid previous ball position";
+      }
+      this._placeInHash(currentHashPos, ball);
+    }
+  }
+
+  /**
+   * Places a ball in spatial hash.
+   * @param {p5.Vector} hashPos
+   * @param {Ball} ball
+   * @private
+   */
+  _placeInHash(hashPos, ball) {
+    const hashKey = hashPos.x + "-" + hashPos.y;
+
+    if (!this._spatialHash[hashKey]) this._spatialHash[hashKey] = [];
+    this._spatialHash[hashKey].push(ball);
+  }
+
+  /**
+   * Removes a ball from the spatial hash.
+   * @param {p5.Vector} hashPos
+   * @param {Ball} ball
+   * @return {boolean}
+   * @private
+   */
+  _removeFromHash(hashPos, ball) {
+    const hashKey = hashPos.x + "-" + hashPos.y;
+
+    if (this._spatialHash[hashKey]) {
+      const ballIndex = this._spatialHash[hashKey].indexOf(ball);
+      if (ballIndex !== -1) {
+        this._spatialHash[hashKey].splice(ballIndex, 1);
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Gets the spatial hashing position of a ball
+   * @param {p5.Vector} pos
+   * @return {p5.Vector}
+   * @private
+   */
+  _getSpatialHashPos(pos) {
```

Comment:
You should rename this to something like `_floorCoordinatesToNearestStaticDiameter`. That will make it more intuitive what this function is doing.

Reply:
That's a good point. I removed the spatial hashing but I will keep this in mind.